cpu/stm32/bootloader.c

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 2020 Benjamin Valentin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     cpu_stm32
+ * @{
+ *
+ * @file
+ * @brief       Trigger reset to the bootloader stored in the internal boot ROM
+ *              memory.
+ *
+ *              This will start the DFU/UART/SPI bootloader.
+ *              See application note AN2606 for which options are available on
+ *              your individual MCU.
+ *
+ * @author      Benjamin Valentin <benpicco@googlemail.com>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "periph_cpu.h"
+
+#define BOOTLOADER_MAGIC    0xB007AFFE
+
+static uint32_t _magic __attribute__((section(".noinit")));
```

Comment:
I wonder if it starts to make sense to have a `magic` module that could be reused for detecting soft reboots and passing flags / parameters over soft reboots. Often hardware even has special registers that allow to reliably detect soft reboots, rather than hoping that uninitialized SRAM does not have by chance the value of a magic cookie. With this being in a dedicated module, it would be easy to have the architecture specific code setting the cookie to zero on a cold boot to increase the chance of that location in SRAM not randomly having the value of the magic cookie to 100.0%. And without architectural support, the cookie approach is still a good best effort fallback.

But the main advantage I see is that this would allow to use less RAM for this, if multiple modules need to pass info beyond a soft reboot.

```C
enum {
    MAGIC_KEY_ENTER_BOOTLOADER, /**< Reboot requests entering the bootloader */
    MAGIC_KEY_NUMOF, /**< Number of Keys */
};

enum {
    MAGIC_FLAG_PUF_SRAM 0x01,
};

typedef struct {
    uint16_t cookie; /**< Used to detect a cold boot (uninitialized RAM) */
    uint8_t key; /**< Single magic key to test for */
    uint8_t flags; /**< Flags to test for */
} boot_magic_t;
```

Reply:
E.g. for PUF_SRAM this would also be useful.

Reply:
Or maybe better provide a function, that could be replaced with architecture specific code if available?

```C
bool magic_is_sram_uninitialized __attribute__((weak)) {
    const uint32_t cookie = 0x52494f54; /* <-- ASCII of "RIOT" */
    static uint32_t magic __attribute__((section(".noinit")));
    if (magic != cookie) {
        magic = cookie;
        return true;
    }
    return false;
}
```

Reply:
I like the idea, because this is something Applications might use too. But I'd view that as orthogonal to this PR.

Since here we only want to signal from shortly before reset to right after reset (even before memory has been initialized), we could save the `static` variable entirely and just write to a random but well-known address.

That's what `samd21-arduino-bootloader` does anyway (writes to the last word of memory).